### PR TITLE
Fix for BRIDGE-3288

### PR DESCRIPTION
**Pull request by alxdarksage**
Once users start creating events before the start of the study, we're in uncharted territory. The fix here is to use negative weeks, and recognize that the day count is running the other direction week after week (because it's negative).

### CODE REVIEW

File: src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java [hunk 1]
```diff
@@ -135,11 +110,48 @@ public StudyAdherenceReport generate(AdherenceState state) {
...
+        if (!ParticipantStudyProgress.NO_ADHERENCE.contains(progression)) {
+            adherence = calculateAdherencePercentage(ImmutableList.of(studyStream));
+        }
+        // If the earliest date is before the study start date, we're going to offset the weekInStudy 
+        // field to show there are scheduled activities occurring earlier in the study design.
+        int weekOffset = 0;
+        if (earliestDate != null && studyStartDate != null && earliestDate.isBefore(studyStartDate)) {
+            int days = Days.daysBetween(earliestDate, studyStartDate).getDays();
+            weekOffset = days/7;
+        }
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java:150]: Off-by-one error: take the example of a day or two before the study, this this is Week 0, is it not?

File: src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java [hunk 2]
```diff
@@ -240,6 +252,11 @@ protected void calculateRowsAndLabels(StudyReportWeek oneWeek, LocalDate localTo
             oneWeek.getByDayEntries().put(i, paddedDays);
         }
         oneWeek.getSearchableLabels().addAll(labels);
+        // In recalculating the rows on the second pass for carry-overs, those carry-overs may 
+        // have row descriptors that differ only in their week, and that means in this new week 
+        // they look identical to the row descriptor for this current week. So we need to reset 
+        // the rows that have been calculated here. 
+        oneWeek.getRows().clear();
         oneWeek.getRows().addAll(rowList);
```
Comment (alxdarksage) [inline src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java]: Revert this, you did it in the wrong PR.

File: src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java [hunk 3]
```diff
@@ -1359,10 +1369,298 @@ public void dateRangeErrorDefaultsToStreamDateRange() {
...
         StudyAdherenceReport report = INSTANCE.generate(state);
-        assertNull(report.getDateRange());
+        assertEquals(report.getDateRange().getStartDate(), LocalDate.parse("2022-02-08"));
+        assertEquals(report.getDateRange().getEndDate(), LocalDate.parse("2022-03-08"));
+        
+        try {
+            System.out.println(BridgeObjectMapper.get().writeValueAsString(report));
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+        }
```
Comment (alxdarksage) [inline src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java]: Remove this

File: src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java [hunk 4]
```diff
@@ -1359,10 +1369,298 @@ public void dateRangeErrorDefaultsToStreamDateRange() {
...
+        assertStartDate(weeklyReport, 0, LocalDate.parse("2022-04-08"));
+        assertStartDate(weeklyReport, 1, LocalDate.parse("2022-04-09"));
+        assertStartDate(weeklyReport, 2, LocalDate.parse("2022-04-10"));
+        assertStartDate(weeklyReport, 3, LocalDate.parse("2022-04-11"));
+        assertStartDate(weeklyReport, 4, LocalDate.parse("2022-04-12"));
+        assertStartDate(weeklyReport, 5, LocalDate.parse("2022-04-13"));
+        assertStartDate(weeklyReport, 6, LocalDate.parse("2022-04-14"));
+    }
+    
+    // BRIDGE-3288 + BRIDGE-3287
```
Comment (alxdarksage) [inline src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java]: Change comment

File: src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java [hunk 5]
```diff
@@ -1359,10 +1369,298 @@ public void dateRangeErrorDefaultsToStreamDateRange() {
...
+                .withMetadata(timeline.getMetadata())
+                .withEvents(ImmutableList.of(event1, event2))
+                .withNow(DateTime.parse("2022-04-24T12:00:00.000Z"))
+                .withStudyStartEventId("custom:event2")
+                .build();
+            
+        StudyAdherenceReport report = StudyAdherenceReportGenerator.INSTANCE.generate(state);
+        
+        StudyReportWeek week = report.getWeeks().get(0);
+        assertEquals(week.getWeekInStudy(), -1);
```
Comment (alxdarksage) [inline src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java:1614]: This is the off-by-one error, shouldn't this be Week -2?

File: src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java (hunk 1 above)
Comment (alxdarksage) [reply src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java:150]: It is actually okay as the earliestDate and studyStartDate are in the same week. It just changes where the two are in relation to one another in the week. It's only once the offset is 7+ that we need to create "negative" weeks (0, -1, etc.).

File: src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java (hunk 4 above)
Comment (alxdarksage) [reply src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java]: Changed

File: src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java (hunk 3 above)
Comment (alxdarksage) [reply src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java]: Removed

File: src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java (hunk 5 above)
Comment (alxdarksage) [reply src/test/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGeneratorTest.java:1614]: No. Notice that the study start event on 4/20, is correctly located in Week 1. We work backwards to week -1 which correctly starts with 4/5, the earliest event.

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**